api/src/main/java/org/apache/iceberg/types/Comparators.java

```diff
@@ -187,15 +195,18 @@ private NullsFirst() {
 
     @Override
     public int compare(T o1, T o2) {
+      if (o1 == o2) {
+        return 0;
+      }
+
       if (o1 != null) {
         if (o2 != null) {
           return 0;
         }
         return 1;
-      } else if (o2 != null) {
-        return -1;
       }
-      return 0;
+
+      return -1;
```

Comment:
The `o1 == o2` case removes the possibility that both are null, so `o2` must be non-null here.